### DISCUSSION TITLE
3D Classification Task

### DISCUSSION THREAD

**Discussion by Meddebma** [Project-MONAI/MONAI#2681] [Q&A]
Hi, 

I am trying to make a 3D classification Task with Images that have a mask and a label. Therefore I used the ImageDataset passing the images, segmentation masks and labels in it. My Idea is that the classification should not be done for the whole image, but only the part of the image that has been selected (In this case it is the spleen). For a better understanding I have two diseases of the spleen in CT scans, and I would like to make a classifier to differentiate between them. 

Here is my Notebook: 
https://github.com/Meddebma/pyradiomics/blob/master/3D_Classification.ipynb

my Questions:
1) When I use the ImageDataset, should I use a different transform for the segmentation mask?
2) How should the original image be cropped so that only the spleen is gonna be isolated for the classification?
3) Here is the Error I get during training:
Expected object of scalar type Long but got scalar type Float for argument #2 'target' in call to _thnn_nll_loss_forward

Thank you

**Comment by Nic-Ma** [comment]
Hi @Meddebma ,

I thnk maybe you can use `MaskIntensityd` transform to mask the spleen region:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/intensity/dictionary.py#L743
Maybe @rijobro @ericspod can share more research approaches to train the model.

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma, Hi @rijobro 
Thanks for your help, I did try to implement MaskIntensityd in that way:

train_transforms = Compose([ScaleIntensity(), AddChannel(), Resize(
    (96, 96, 96)), ToTensor()])
val_transforms = Compose(
    [ScaleIntensity(), AddChannel(), Resize((96, 96, 96)), ToTensor()])
seg_transforms = Compose([ScaleIntensity(), AddChannel(), Resize(
    (96, 96, 96)), **MaskIntensity(mask_data=1)**, ToTensor()])

 create a training data loader
train_ds = ImageDataset(
    image_files=images[:18], seg_files=segmentations[:18], labels=labels[:18], transform=train_transforms, seg_transform=seg_transforms)
train_loader = DataLoader(train_ds, batch_size=1, shuffle=True,
                          num_workers=4, pin_memory=torch.cuda.is_available())

Unfortunately I've got this error:
 if mask_data_.shape[0] != 1 and mask_data_.shape[0] != img.shape[0]:
IndexError: tuple index out of range

How can I fix it? Thanks a lot 
Here is the notebook:
https://github.com/Meddebma/pyradiomics/blob/master/3D_Classification.ipynb

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

Thanks for providing the debug info.
I think there are 3 main things need to update for your case:
1. As you are working with both `label` and `mask`, suggest to use `dictionary` based data and transforms.
2. Need to `AddChannel` first before other transforms.
3. `MaskIntensity` require numpy array input instead of int number.

Here is a proposal transforms:
```py
data = [{"image": i, "seg": s, "label": l} for i, s, l in zip(images, segmentations, labels)]

transforms = Compose([
    LoadImaged(keys=["image", "seg"]),
    EnsureChannelFirstd(keys=["image", "seg"]),
    ScaleIntensityd(keys="image"),
    Resized(keys=["image", "seg"], spatial_size=(96, 96, 96)),
    MaskIntensityd(keys="image", mask_key="seg"),
    ToTensor(keys=["image", "label"]),
])

dataset = Dataset(data=data, transforms=transforms)
```
The source code of `MaskIntensity` is a little bit confusing, let me try to make it more clear later.

Thanks.

**Reply by Nic-Ma** [reply]
Submitted PR: https://github.com/Project-MONAI/MONAI/pull/2726 to optimize the `MaskIntensity` transform.

Thanks.

**Reply by Meddebma** [reply]
@Nic-Ma 
Thanks a lot, I did some modifications and I've got quite good results using this code:
`data = [{"image": i, "seg": s, "label": l} for i, s, l in zip(images, segmentations, labels)]

transforms = Compose([
    LoadImaged(keys=["image", "seg"]),
    AddChanneld(keys=["image", "seg"]),
    Resized(keys=["image", "seg"], spatial_size=(96, 96, 96)),
    ScaleIntensityRanged(keys=['image'], a_min=-57, a_max=164, b_min=0.0, b_max=1.0, clip=True),
    MaskIntensityd(keys="image", mask_key="seg"),
    ToTensord(keys=["image", "seg"]),
])

train_ds = CacheDataset(data=data[:10], transform=transforms)
train_loader = DataLoader(train_ds, batch_size=1, shuffle=True,
                          num_workers=4, pin_memory=torch.cuda.is_available())`

Unfortunately the edges of the image are not smooth as in the label as you can see it in the notebook:
https://github.com/Meddebma/pyradiomics/blob/master/3D_Classification.ipynb

can I somehow fix it?
Thanks

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

That may because the default interpolation mode of `Resized` transform can't work well.
You can change to other modes:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/spatial/dictionary.py#L511
For example:
```py
Resized(keys=["image", "seg"], spatial_size=(96, 96, 96), mode=["bilinear", "nearest"])
```

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma,

So I followed your advice, I've got this error after changing the modes:
NotImplementedError: Got 5D input, but bilinear mode needs 4D input

Same problem happened with the spacing transform actually.

Here is the notebook 
https://github.com/Meddebma/pyradiomics/blob/master/3D_Classification.ipynb

thanks

**Reply by Nic-Ma** [reply]
Seems `bilinear` can't work with your data dims, could you please help try other modes? Like "linear", etc.

**Reply by Meddebma** [reply]
Hi @Nic-Ma, 
thanks, it seems that also other modes dont work, but its fine, I will just use Rezised for training as it is.

During the Training I've got a problem during the validation:
**AssertionError: y values must be 0 or 1, can not be all 0 or all 1.**

here  is the notebook:
https://github.com/Meddebma/pyradiomics/blob/master/3D_Classification.ipynb

thanks!

**Reply by Nic-Ma** [reply]
That because you only used 2 images for validation, and the labels are `1`, which is not allowed in AUC metric.
AUC can't work with all 0 or all 1 labels in theory.

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma,

Ah I see, is there a possibility to randomly make 80% of the data for training and 20% for validation instead of using data[:18]?

Thanks

**Reply by Nic-Ma** [reply]
You can take a look at the `partition_dataset` API:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/utils.py#L816

Thanks.

**Reply by Meddebma** [reply]
I tried this:
data= [{"image": i, "seg": s, "label": l} for i, s, l in zip(images, segmentations, labels)]
data_partition= partition_dataset_classes(images, labels, shuffle=True, ratios=[0.8, 0.2])
and I got this error:
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

here is the notebook:
https://github.com/Meddebma/Complete-Python-3-Bootcamp/blob/master/3D_Classification.ipynb

Thanks

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

I think the `classes` arg is not flexible enough to handle all kinds of input, could you please help try below code:
```py
data_partition= partition_dataset_classes(data, labels.tolist(), shuffle=True, ratios=[0.8, 0.2])
```
I will submit a quick PR to enhance the API.

Thanks.

**Reply by Nic-Ma** [reply]
Submitted PR: https://github.com/Project-MONAI/MONAI/pull/2761 for it.

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma,

I tried the code and I've got only 2 images in the trainloader as you can see in the notebook:
https://github.com/Meddebma/Complete-Python-3-Bootcamp/blob/master/3D_Classification.ipynb

Did I implement it wrong in the train and validation dataloaders?

thanks

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

That because the returned data of `partition_dataset` is 2 items: `train_data` and `val_data` according to your ratio.
For example:
```
        >>> partition_dataset_classes(data, classes, shuffle=False, ratios=[2, 1])
        [[2, 8, 4, 1, 3, 6, 5, 11, 12], [10, 13, 7, 9, 14]]
```

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma,

thanks for helping out, when I run the cell as it is the system crashes after 10min and restarts again, do you think I made a Mistake in my code?

thanks

**Reply by Nic-Ma** [reply]
Could you please help share the error log?

Thanks.

**Reply by Meddebma** [reply]
There was unfortunately no error log, after circa 10min the RAM crashed and the session restarted and the cell output showed:
Loading dataset:   0%|          | 0/2 [00:00<?, ?it/s]
is it maybe because I used CacheDataset? 

Thanks

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

As I said in the previous thread: the returned data of partition_dataset is 2 items: train_data and val_data, so I think your `CacheDataset` can set as:
```py
datasets = partition_dataset_classes(data, classes, shuffle=False, ratios=[2, 1])
train_ds = CacheDataset(data=datasets[0], ...)
val_ds = CacheDataset(data=datasets[1], ...)
```

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma,

it worked perfectly, and even the accuracy went up to 0.96 during the training !!!

Thank you so much for your effort and help!

**Reply by Nic-Ma** [reply]
No problem, glad to see your great progress!
Please feel free to raise a ticket if you face any other questions.

Thanks.
